csrc/python_frontend/translation.cpp

```diff
@@ -0,0 +1,560 @@
+      return;
+    }
+    createScalar(v);
+  }
+
+  // Add Tensor value to Fusion Definition
+  void handle(const TensorView* tv) final {
+    // short-circuit: value already exists in FusionDefinition
+    if (map_val_to_fd_index_.count(tv) > 0) {
+      return;
+    }
+
+    Tensor output = fd_->defineTensor(tv->nDims());
+    map_val_to_fd_index_.emplace(tv, output());
+
+    std::vector<int64_t> shape;
+    std::transform(
+        tv->domain()->logical().begin(),
+        tv->domain()->logical().end(),
+        std::back_inserter(shape),
+        [](IterDomain* id) {
+          return (id->getMaybeExpandedExtent()->isConstScalar())
+              ? id->getMaybeExpandedExtent()->evaluate().as<int64_t>()
+              : -1;
+        });
+
+    fd_->defineRecord(new TensorRecord(
+        {fd_->recordingState(output())},
+        shape,
+        tv->domain()->contiguity(),
+        std::get<PrimDataType>(tv->dtype().type),
+        tv->isCpuScalar(),
+        tv->domain()->strideOrder()));
+  }
+
+  // =================================================================================
+  // Handle add_output variants
+
+  // Add Tensor output to FusionDefinition
+  void handleOutput(const TensorView* tv) {
+    size_t output_index = map_val_to_fd_index_.at(tv);
+    fd_->defineRecord(new OutputRecord<TensorView>(
+        {fd_->recordingState(output_index)},
+        serde::RecordType::OutputTv,
+        tv->domain()->strideOrder()));
+  }
+
+  // Alias output Tensor with input tensor
+  void handleOutput(const TensorView* tv, const AliasInfo& alias_info) {
+    size_t output_index = map_val_to_fd_index_.at(tv);
+    size_t input_index = map_val_to_fd_index_.at(alias_info.aliased_io);
+    fd_->defineRecord(new OutputRecord<TensorView>(
+        {fd_->recordingState(output_index), fd_->recordingState(input_index)},
+        serde::RecordType::OutputTv));
+  }
+
+  // =================================================================================
+  // Map CPP Expression classes to corresponding RecordFunctors in
+  // python_frontend
+
+  // A generic function to map UnaryOp, BinaryOp, and TernaryOp to
+  // python_frontend OpRecord
+  template <typename ExprType, typename ResultType, typename... ArgTypes>
+  void handleOpRecord(
+      const Expr* e,
+      serde::RecordType record_type,
+      ResultType result,
+      ArgTypes... args) {
+    NVF_ERROR(e->isA<ExprType>());
+    std::vector<State> argument_states;
+    std::transform(
+        e->inputs().begin(),
+        e->inputs().end(),
+        std::back_inserter(argument_states),
+        [&](auto arg) {
+          return fd_->recordingState(map_val_to_fd_index_.at(arg));
+        });
+
+    fd_->defineRecord(new OpRecord<ResultType, ArgTypes...>(
+        argument_states,
+        {fd_->recordingState(map_val_to_fd_index_.at(result))},
+        "ops." + getString(e->as<ExprType>()),
+        record_type,
+        getFunction<ResultType, ArgTypes...>(e->as<ExprType>())));
```

Comment:
If I'm being a language police, I would say that the template `getFunction` seems a bit too complicated. I think you can construct a function type here with `ResultType(*)(ArgTypes...)` and use static_cast along to resolve the overload.

Reply:
Returning a function pointer has ugly syntax, so I went with returning `std::function`.